gdx/src/com/badlogic/gdx/utils/ObjectMap.java

```diff
@@ -1,846 +1,782 @@
+	 * work in a hash table without heavy collision rates. It has known problems when all or most hashCode()s are
+	 * multiples of larger Fibonacci numbers; see <a href="https://probablydance.com/2018/06/16/fibonacci-hashing-the-optimization-that-the-world-forgot-or-a-better-alternative-to-integer-modulo/">this blog post by Malte Skarupke</a>
+	 * for more details. In the unlikely event that most of your hashCode()s are Fibonacci numbers, you can subclass
+	 * this to change this method, which is a one-liner in this form:
+	 * {@code return (int) (item.hashCode() * 0x9E3779B97F4A7C15L >>> shift);}
+	 * <br>
+	 * This can be overridden by subclasses, which you may want to do if your key type needs special consideration for
+	 * its hash (such as if you can't modify or extend a particular class that has an incorrect hashCode()). Subclasses
+	 * that don't need the collision decrease of Fibonacci Hashing (assuming the key class has a good hashCode()) may do
+	 * fine with a simple implementation:
+	 * {@code return (item.hashCode() & mask);}
+	 *
+	 * @param item a key that this method will hash, by default by calling {@link Object#hashCode()} on it; non-null
+	 * @return an int between 0 and {@link #mask}, both inclusive
+	 */
+	protected int place (final K item) {
+		// shift is always greater than 32, less than 64
+		return (int)(item.hashCode() * 0x9E3779B97F4A7C15L >>> shift);
+	}
+
+	private int locateKey (final K key) {
+		return locateKey(key, place(key));
+	}
+
+	/**
+	 * Given a key and its initial placement to try in an array, this finds the actual location of the key in the array
+	 * if it is present, or -1 if the key is not present. This can be overridden if a subclass needs to compare for
+	 * equality differently than just by calling {@link Object#equals(Object)}, but only within the same package.
+	 *
+	 * @param key       a K key that will be checked for equality if a similar-seeming key is found
+	 * @param placement as calculated by {@link #place(Object)}, almost always with {@code place(key)}
+	 * @return the location in the key array of key, if found, or -1 if it was not found.
+	 */
+	int locateKey (final K key, final int placement) {
+		for (int i = placement; ; i = i + 1 & mask) {
+			// empty space is available
+			if (keyTable[i] == null) {
+				return -1;
+			}
+			if (key.equals(keyTable[i])) {
+				return i;
+			}
+		}
+	}
+
+	/**
+	 * Returns the old value associated with the specified key, or null.
+	 */
+	public V put (K key, V value) {
+		if (key == null)
+			throw new IllegalArgumentException("key cannot be null.");
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		int b = place(key);
+		int loc = locateKey(key, b);
+		// an identical key already exists
+		if (loc != -1) {
+			V tv = valueTable[loc];
+			valueTable[loc] = value;
+			return tv;
+		}
+		for (int i = b; ; i = (i + 1) & mask) {
+			// space is available so we insert and break (resize is later)
+			if (keyTable[i] == null) {
```

Comment:
`locateKey` has already done the work to find the next empty slot, it's wasteful to find it again.

Reply:
The highlighted line 200 only runs if `loc == -1`, meaning the given key was not found already present in the map. `locateKey()` could be changed, or a variant added, that somehow can give either the position of the first empty space or the position of the found key. I have some idea of how I could do it.

Reply:
`locateKey` could be inlined for put, so the code can handle an existing key and still make use of the location of the next null. Java's lack of tuples make using a method for this awkward.